In [ ]:
#importing the necessary libraries
import numpy as np
import pandas as pd
import seaborn as sns
from sklearn.linear_model import LinearRegression  # for Linear regression
from sklearn.model_selection import train_test_split
from sklearn import metrics
import matplotlib.pyplot as plt



In [ ]:
hp=pd.read_csv('HousePrices.csv') # importing the dataset
hp.head()

Id  Dwell_Type Zone_Class  LotFrontage  LotArea Road_Type Alley  \
0   1          60         RL         65.0     8450      Pave   NaN   
1   2          20         RL         80.0     9600      Pave   NaN   
2   3          60         RL         68.0    11250      Pave   NaN   
3   4          70         RL         60.0     9550      Pave   NaN   
4   5          60         RL         84.0    14260      Pave   NaN   

  Property_Shape LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature  \
0            Reg         Lvl    AllPub  ...        0    NaN   NaN         NaN   
1            Reg         Lvl    AllPub  ...        0    NaN   NaN         NaN   
2            IR1         Lvl    AllPub  ...        0    NaN   NaN         NaN   
3            IR1         Lvl    AllPub  ...        0    NaN   NaN         NaN   
4            IR1         Lvl    AllPub  ...        0    NaN   NaN         NaN   

  MiscVal MoSold YrSold  SaleType  SaleCondition  Property_Sale_Price  
0       0      2   2008        WD         Normal               208500  
1       0      5   2007        WD         Normal               181500  
2       0      9   2008        WD         Normal               223500  
3       0      2   2006        WD        Abnorml               140000  
4       0     12   2008        WD         Normal               250000  

[5 rows x 81 columns]

### Basic data Exploration 

In [ ]:
hp.shape

(2073, 81)

In [ ]:
hp.describe().T

count           mean           std      min       25%  \
Id                   2073.0     916.132176    493.014670      1.0     519.0   
Dwell_Type           2073.0      60.556199    159.924810     20.0      20.0   
LotFrontage          1753.0      72.056475     27.987475     21.0      60.0   
LotArea              2073.0   10717.853353   9215.982306   1300.0    7620.0   
OverallQual          2073.0       6.233478      1.517115      1.0       5.0   
OverallCond          2073.0       5.835022      1.360214      1.0       5.0   
YearBuilt            2073.0    1969.445731     30.222647   1872.0    1950.0   
YearRemodAdd         2073.0    1986.617463     20.432694   1950.0    1970.0   
MasVnrArea           2059.0     101.875668    179.169217      0.0       0.0   
BsmtFinSF1           2073.0     437.949349    448.051119      0.0       0.0   
BsmtFinSF2           2073.0      49.219489    165.621465      0.0       0.0   
BsmtUnfSF            2073.0     565.101302    449.730940      0.0     208.0   
TotalBsmtSF          2073.0    1052.270140    431.026436      0.0     794.0   
1stFlrSF             2073.0    1166.096961    376.975119    334.0     894.0   
2ndFlrSF             2073.0     344.745779    438.844748      0.0       0.0   
LowQualFinSF         2073.0       6.168355     49.553967      0.0       0.0   
GrLivArea            2073.0    1517.011095    517.236428    334.0    1126.0   
BsmtFullBath         2073.0       0.429329      0.522605      0.0       0.0   
BsmtHalfBath         2073.0       0.056440      0.237015      0.0       0.0   
FullBath             2073.0       1.571635      0.545963      0.0       1.0   
HalfBath             2073.0       0.378678      0.499873      0.0       0.0   
BedroomAbvGr         2073.0       2.874096      0.816144      0.0       2.0   
KitchenAbvGr         2073.0       1.052581      0.237901      0.0       1.0   
TotRmsAbvGrd         2073.0       6.540280      1.629673      2.0       5.0   
Fireplaces           2073.0       0.602026      0.633931      0.0       0.0   
GarageYrBlt          1960.0    1980.418367     24.158327   1906.0    1963.0   
GarageCars           2073.0       1.775687      0.741976      0.0       1.0   
GarageArea           2073.0     474.683068    212.284955      0.0     336.0   
WoodDeckSF           2073.0      94.168837    129.072113      0.0       0.0   
OpenPorchSF          2073.0      46.035697     64.768713      0.0       0.0   
EnclosedPorch        2073.0      21.029908     59.569928      0.0       0.0   
3SsnPorch            2073.0       2.979257     26.817218      0.0       0.0   
ScreenPorch          2073.0      14.552340     54.099626      0.0       0.0   
PoolArea             2073.0       2.299083     37.411031      0.0       0.0   
MiscVal              2073.0      39.736614    429.819512      0.0       0.0   
MoSold               2073.0       6.287506      2.690130      1.0       5.0   
YrSold               2073.0    2007.851905      1.329705   2006.0    2007.0   
Property_Sale_Price  2073.0  180967.284129  78666.411112  34900.0  130000.0   

                          50%       75%       max  
Id                      932.0    1302.0    1820.0  
Dwell_Type               50.0      70.0    7080.0  
LotFrontage              70.0      83.0     313.0  
LotArea                9492.0   11601.0  215245.0  
OverallQual               6.0       7.0      10.0  
OverallCond               5.0       7.0       9.0  
YearBuilt              1971.0    1999.0    2010.0  
YearRemodAdd           1995.0    2004.0    2023.0  
MasVnrArea                0.0     161.5    1600.0  
BsmtFinSF1              381.0     708.0    5644.0  
BsmtFinSF2                0.0       0.0    1474.0  
BsmtUnfSF               466.0     804.0    2336.0  
TotalBsmtSF             990.0    1291.0    6110.0  
1stFlrSF               1095.0    1391.0    4692.0  
2ndFlrSF                  0.0     730.0    2065.0  
LowQualFinSF              0.0       0.0     572.0  
GrLivArea              1470.0    1792.0    5642.0  
Bsmt

In [ ]:
hp.isnull().sum()

Id                       0
Dwell_Type               0
Zone_Class               0
LotFrontage            320
LotArea                  0
                      ... 
MoSold                   0
YrSold                   0
SaleType                 0
SaleCondition            0
Property_Sale_Price      0
Length: 81, dtype: int64

In [ ]:
hp.duplicated(subset=None, keep='first') #checking for duplicates 

0       False
1       False
2       False
3       False
4       False
        ...  
2068    False
2069    False
2070    False
2071    False
2072    False
Length: 2073, dtype: bool

In [ ]:
hp.drop_duplicates(inplace=True) # droping the duplicates

In [ ]:
hp.shape  # 13 rows had been dropped

(2060, 81)

In [ ]:
hp.shape

(2060, 81)

In [ ]:
hp.corr()

Id  Dwell_Type  LotFrontage   LotArea  OverallQual  \
Id                   1.000000    0.043299     0.067185  0.004239     0.083396   
Dwell_Type           0.043299    1.000000    -0.075498 -0.033410     0.046145   
LotFrontage          0.067185   -0.075498     1.000000  0.310674     0.108867   
LotArea              0.004239   -0.033410     0.310674  1.000000     0.078304   
OverallQual          0.083396    0.046145     0.108867  0.078304     1.000000   
OverallCond          0.195934    0.033409    -0.010685  0.013932    -0.009635   
YearBuilt           -0.080607    0.026858     0.092788 -0.009935     0.334836   
YearRemodAdd         0.087593    0.042514     0.101759  0.017978     0.377353   
MasVnrArea          -0.027116   -0.005125     0.130239  0.063852     0.284305   
BsmtFinSF1          -0.016479   -0.033724     0.141027  0.164464     0.164686   
BsmtFinSF2           0.005005   -0.025552     0.062477  0.075633    -0.058468   
BsmtUnfSF           -0.007185   -0.022291     0.111681 -0.000064     0.197035   
TotalBsmtSF         -0.022718   -0.068189     0.285726  0.200109     0.354552   
1stFlrSF             0.008018   -0.078211     0.359419  0.244565     0.309609   
2ndFlrSF             0.002789    0.100750     0.068945  0.036910     0.209747   
LowQualFinSF        -0.023792    0.004968     0.023330 -0.001376    -0.013781   
GrLivArea            0.005923    0.028839     0.320977  0.209383     0.402039   
BsmtFullBath         0.021072   -0.015975     0.087816  0.125910     0.082388   
BsmtHalfBath        -0.030907   -0.004557    -0.038047  0.031376    -0.048188   
FullBath             0.023644    0.055960     0.176575  0.098769     0.382172   
HalfBath             0.000197    0.065944     0.047937  0.004216     0.188945   
BedroomAbvGr         0.011195   -0.005557     0.228040  0.107632     0.062114   
KitchenAbvGr         0.007700    0.074569     0.005882 -0.007142    -0.108729   
TotRmsAbvGrd         0.009741    0.019239     0.293949  0.155715     0.275057   
Fireplaces          -0.010898   -0.031859     0.225763  0.210605     0.253350   
GarageYrBlt          0.075136    0.043836     0.067632 -0.021976     0.379662   
GarageCars           0.028357   -0.002059     0.234399  0.129533     0.399316   
GarageArea           0.016723   -0.014408     0.271025  0.145898     0.373199   
WoodDeckSF           0.006050    0.004788     0.082239  0.120960     0.154052   
OpenPorchSF         -0.005012    0.001076     0.110677  0.065463     0.215683   
EnclosedPorch        0.001390   -0.006085    -0.001902 -0.023194    -0.079744   
3SsnPorch           -0.045437   -0.011985     0.043145  0.010444     0.000130   
ScreenPorch         -0.004662   -0.011871     0.022591  0.035586     0.021721   
PoolArea             0.031310    0.001617     0.124490  0.062409     0.045044   
MiscVal             -0.007308   -0.003824    -0.003727  0.031199    -0.033984   
MoSold               0.000152   -0.020480     0.026055 -0.000814     0.032628   
YrSold               0.043669    0.018580     0.023621 -0.007590    -0.018586   
Property_Sale_Price -0.002564   -0.017638     0.280675  0.211359     0.522678   

                     OverallCond  YearBuilt  YearRemodAdd  MasVnrArea  \
Id                      0.195934  -0.080607      0.087593   -0.027116   
Dwell_Type              0.033409   0.026858      0.042514   -0.005125   
LotFrontage            -0.010685   0.092788      0.101759    0.130239   
LotArea                 0.013932  -0.009935      0.017978    0.063852   
OverallQual            -0.009635   0.334836      0.377353    0.284305   
OverallCond             1.000000  -0.243986      0.085939   -0.083318   
YearBuilt              -0.243986   1.000000      0.534376    0.265946   
YearRemodAdd            0.085939   0.534376      1.000000    0.167786   
MasVnrArea             -0.083318   0.265946      0.167786    1.000000   
BsmtFinSF1             -0.037001   0.210807      0.103886    0.238482   
BsmtFinSF2              0.027185  -0.026433     -0.073611   -0.0

In [ ]:
#dropping unnecessary columns
hp.drop(columns=['Dwell_Type','OverallCond'],inplace=True)
hp.drop(columns=['LowQualFinSF','BsmtHalfBath'],inplace=True)
hp.drop(columns=['MoSold','YrSold'],inplace=True)
hp.drop(columns=['Alley'],inplace=True)
hp.drop('PoolQC', inplace=True, axis=1)
hp.drop('MiscFeature', inplace=True, axis=1)
hp.drop(columns=['Fence'],inplace=True)

In [ ]:
hp['LotFrontage']

0        65.0
1        80.0
2        68.0
3        60.0
4        84.0
        ...  
2068      NaN
2069     63.0
2070    313.0
2071     52.0
2072     98.0
Name: LotFrontage, Length: 2060, dtype: float64

##Replacing Missing values on numerical attributes

In [ ]:
#replacing LotFrontange column with mean value
mean_value=hp['LotFrontage'].mean()
hp['LotFrontage'].fillna(value=mean_value,inplace=True)

In [ ]:
hp['LotFrontage'].isnull().sum()

0

In [ ]:
mean_value=hp['MasVnrArea'].mean()
hp['MasVnrArea'].fillna(value=mean_value,inplace=True)

In [ ]:
hp['MasVnrArea'].isnull().sum()

0

In [ ]:
mean_value=hp['GarageYrBlt'].mean()
hp['GarageYrBlt'].fillna(value=mean_value,inplace=True)

###Replacing missing values on categorical datatype

In [ ]:
hp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2060 entries, 0 to 2072
Data columns (total 71 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   2060 non-null   int64  
 1   Zone_Class           2060 non-null   object 
 2   LotFrontage          2060 non-null   float64
 3   LotArea              2060 non-null   int64  
 4   Road_Type            2060 non-null   object 
 5   Property_Shape       2060 non-null   object 
 6   LandContour          2060 non-null   object 
 7   Utilities            2060 non-null   object 
 8   LotConfig            2060 non-null   object 
 9   LandSlope            2060 non-null   object 
 10  Neighborhood         2060 non-null   object 
 11  Condition1           2060 non-null   object 
 12  Condition2           2060 non-null   object 
 13  Dwelling_Type        2060 non-null   object 
 14  HouseStyle           2060 non-null   object 
 15  OverallQual          2060 non-null   i

In [ ]:
mode_value=hp['MasVnrType'].mode()
hp['MasVnrType'].fillna(value=mode_value,inplace=True)

In [ ]:
hp['MasVnrArea'].isnull().sum()

0

In [ ]:
mode_value=hp['BsmtQual'].mode()
hp['BsmtQual'].fillna(value=mode_value,inplace=True)

In [ ]:
mode_value=hp['BsmtCond'].mode()
hp['BsmtCond'].fillna(value=mode_value,inplace=True)

In [ ]:
mode_value=hp['BsmtExposure'].mode()
hp['BsmtExposure'].fillna(value=mode_value,inplace=True)

In [ ]:
mode_value=hp['BsmtFinType2'].mode()
hp['BsmtFinType2'].fillna(value=mode_value,inplace=True)

In [ ]:
mode_value=hp['BsmtFinType1'].mode()
hp['BsmtFinType1'].fillna(value=mode_value,inplace=True)

In [ ]:
hp=hp.fillna(hp.mode().iloc[0])

In [ ]:
hp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2060 entries, 0 to 2072
Data columns (total 71 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   2060 non-null   int64  
 1   Zone_Class           2060 non-null   object 
 2   LotFrontage          2060 non-null   float64
 3   LotArea              2060 non-null   int64  
 4   Road_Type            2060 non-null   object 
 5   Property_Shape       2060 non-null   object 
 6   LandContour          2060 non-null   object 
 7   Utilities            2060 non-null   object 
 8   LotConfig            2060 non-null   object 
 9   LandSlope            2060 non-null   object 
 10  Neighborhood         2060 non-null   object 
 11  Condition1           2060 non-null   object 
 12  Condition2           2060 non-null   object 
 13  Dwelling_Type        2060 non-null   object 
 14  HouseStyle           2060 non-null   object 
 15  OverallQual          2060 non-null   i

In [ ]:
hp.drop('Road_Type', inplace=True, axis=1)

###Label Encoding

In [ ]:
from sklearn import preprocessing
 
# label_encoder object knows how to understand word labels.
label_encoder=preprocessing.LabelEncoder()
 
# Encode labels in column 'species'.
hp['Zone_Class']= label_encoder.fit_transform(hp['Zone_Class'])

In [ ]:

hp['LandContour']= label_encoder.fit_transform(hp['LandContour'])


In [ ]:
hp['Property_Shape']= label_encoder.fit_transform(hp['Property_Shape'])
hp['Utilities']= label_encoder.fit_transform(hp['Utilities'])
hp['LotConfig']= label_encoder.fit_transform(hp['LotConfig'])
hp['LandSlope']= label_encoder.fit_transform(hp['LandSlope'])
hp['Neighborhood']= label_encoder.fit_transform(hp['Neighborhood'])

In [ ]:
 #dropping the encoded columns  that has less correlations
hp.drop(columns='LandContour',inplace=True)
hp.drop(columns=['Utilities','LotConfig'],inplace=True)

In [ ]:
hp['Condition1']= label_encoder.fit_transform(hp['Condition1'])
hp['Condition2']= label_encoder.fit_transform(hp['Condition2'])
hp['Dwelling_Type']= label_encoder.fit_transform(hp['Dwelling_Type'])
hp['HouseStyle']= label_encoder.fit_transform(hp['HouseStyle'])
hp['RoofStyle']= label_encoder.fit_transform(hp['RoofStyle'])
hp['RoofMatl']= label_encoder.fit_transform(hp['RoofMatl'])
hp['Exterior1st']= label_encoder.fit_transform(hp['Exterior1st'])
hp['Exterior2nd']= label_encoder.fit_transform(hp['Exterior2nd'])
hp['MasVnrType']= label_encoder.fit_transform(hp['MasVnrType'])

In [ ]:
hp['ExterQual']= label_encoder.fit_transform(hp['ExterQual'])
hp['ExterCond']= label_encoder.fit_transform(hp['ExterCond'])
hp['Foundation']= label_encoder.fit_transform(hp['Foundation'])


In [ ]:
hp['BsmtQual']= label_encoder.fit_transform(hp['BsmtQual'])
hp['BsmtCond']= label_encoder.fit_transform(hp['BsmtCond'])
hp['BsmtExposure']= label_encoder.fit_transform(hp['BsmtExposure'])
hp['BsmtFinType1']= label_encoder.fit_transform(hp['BsmtFinType1'])
hp['BsmtFinType2']= label_encoder.fit_transform(hp['BsmtFinType2'])
hp['Heating']= label_encoder.fit_transform(hp['Heating'])
hp['HeatingQC']= label_encoder.fit_transform(hp['HeatingQC'])
hp['CentralAir']= label_encoder.fit_transform(hp['CentralAir'])
hp['Electrical']= label_encoder.fit_transform(hp['Electrical'])
hp['KitchenQual']= label_encoder.fit_transform(hp['KitchenQual'])
hp['Functional']= label_encoder.fit_transform(hp['Functional'])
hp['FireplaceQu']= label_encoder.fit_transform(hp['FireplaceQu'])
hp['GarageType']= label_encoder.fit_transform(hp['GarageType'])


In [ ]:
hp['GarageFinish']= label_encoder.fit_transform(hp['GarageFinish'])
hp['GarageQual']= label_encoder.fit_transform(hp['GarageQual'])
hp['GarageCond']= label_encoder.fit_transform(hp['GarageCond'])
hp['PavedDrive']= label_encoder.fit_transform(hp['PavedDrive'])
hp['SaleType']= label_encoder.fit_transform(hp['SaleType'])
hp['SaleCondition']= label_encoder.fit_transform(hp['SaleCondition'])



In [ ]:
hp.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2060 entries, 0 to 2072
Data columns (total 67 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Id                   2060 non-null   int64  
 1   Zone_Class           2060 non-null   int64  
 2   LotFrontage          2060 non-null   float64
 3   LotArea              2060 non-null   int64  
 4   Property_Shape       2060 non-null   int64  
 5   LandSlope            2060 non-null   int64  
 6   Neighborhood         2060 non-null   int64  
 7   Condition1           2060 non-null   int64  
 8   Condition2           2060 non-null   int64  
 9   Dwelling_Type        2060 non-null   int64  
 10  HouseStyle           2060 non-null   int64  
 11  OverallQual          2060 non-null   int64  
 12  YearBuilt            2060 non-null   int64  
 13  YearRemodAdd         2060 non-null   int64  
 14  RoofStyle            2060 non-null   int64  
 15  RoofMatl             2060 non-null   i

##Building the model 

###1. Linear Regression

In [ ]:
#implementing Linear Regression
X = hp.drop(['Property_Sale_Price'], axis=1)
#Feature selection
# the dependent variable 
y= hp[['Property_Sale_Price']]

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30, random_state=1)

In [ ]:

# fit the model
reg_model = LinearRegression()
reg_model.fit(X_train, y_train)

LinearRegression()

In [ ]:
reg_model.score(X_train, y_train),reg_model.score(X_test, y_test)

(0.8694816615535806, 0.74194207960257)

Here, the training-set accuracy score is **0.8694** while the test-set accuracy to be **0.7419**.
The model is overfitting as in training data the model performs well on the training data but does not perform well on the evaluation data.